acceptance-tests/domain_fronting_test.go

```diff
@@ -41,56 +42,188 @@ var _ = Describe("Domain fronting", func() {
+		haproxyInfo, varsStoreReader = deployHAProxy(baseManifestVars{
 			haproxyBackendPort:    haproxyBackendPort,
 			haproxyBackendServers: []string{"127.0.0.1"},
 			deploymentName:        defaultDeploymentName,
-		}, []string{opsfile}, map[string]interface{}{}, true)
+		}, []string{opsfile}, map[string]interface{}{
+			"disable_domain_fronting": disableDomainFronting,
+			"cert_common_name":        "haproxy.internal",
+			"cert_sans":               []string{"haproxy.internal"},
+		}, true)
 
 		err := varsStoreReader(&creds)
 		Expect(err).NotTo(HaveOccurred())
 
-		closeLocalServer, localPort := startDefaultTestServer()
-		defer closeLocalServer()
+		var localPort int
+		closeLocalServer, localPort = startDefaultTestServer()
+		closeSSHTunnel = setupTunnelFromHaproxyToTestServer(haproxyInfo, haproxyBackendPort, localPort)
 
-		closeTunnel := setupTunnelFromHaproxyToTestServer(haproxyInfo, haproxyBackendPort, localPort)
-		defer closeTunnel()
+		clientCert, err = tls.X509KeyPair([]byte(creds.Client.Certificate), []byte(creds.Client.PrivateKey))
+		Expect(err).NotTo(HaveOccurred())
 
-		httpClient := buildHTTPClient(
+		nonMTLSClient = buildHTTPClient(
 			[]string{creds.HTTPSFrontend.CA},
 			map[string]string{
 				"haproxy.internal:443": fmt.Sprintf("%s:443", haproxyInfo.PublicIP),
 			},
-			[]tls.Certificate{},
+			[]tls.Certificate{}, "",
 		)
 
-		By("Sending a request to HAProxy with a mismatched SNI and Host header it returns a 421")
-		req, err := http.NewRequest("GET", "https://haproxy.internal", nil)
-		Expect(err).NotTo(HaveOccurred())
-		req.Host = "spoof.internal"
+		mtlsClient = buildHTTPClient(
+			[]string{creds.HTTPSFrontend.CA},
+			map[string]string{
+				"haproxy.internal:443": fmt.Sprintf("%s:443", haproxyInfo.PublicIP),
+			},
+			[]tls.Certificate{clientCert}, "",
+		)
 
-		resp, err := httpClient.Do(req)
-		Expect(err).NotTo(HaveOccurred())
-		Expect(resp.StatusCode).To(Equal(http.StatusMisdirectedRequest))
+		nonMTLSClientNoSNI = buildHTTPClient(
+			[]string{creds.HTTPSFrontend.CA},
+			map[string]string{
+				"haproxy.internal:443": fmt.Sprintf("%s:443", haproxyInfo.PublicIP),
+			},
+			[]tls.Certificate{}, "1.2.3.4",
+		)
 
-		By("Sending a request to HAProxy without a spoofed Host header it returns a 200 as normal")
-		resp, err = httpClient.Get("https://haproxy.internal")
-		Expect(err).NotTo(HaveOccurred())
-		Expect(resp.StatusCode).To(Equal(http.StatusOK))
-		Eventually(gbytes.BufferReader(resp.Body)).Should(gbytes.Say("Hello cloud foundry"))
+		mtlsClientNoSNI = buildHTTPClient(
+			[]string{creds.HTTPSFrontend.CA},
+			map[string]string{
+				"haproxy.internal:443": fmt.Sprintf("%s:443", haproxyInfo.PublicIP),
+			},
+			[]tls.Certificate{clientCert}, "1.2.3.4",
+		)
+	})
+
+	AfterEach(func() {
+		if closeLocalServer != nil {
+			defer closeLocalServer()
+		}
+
+		if closeSSHTunnel != nil {
+			defer closeSSHTunnel()
+		}
+	})
+
+	Context("When disable domain fronting is true", func() {
```

Comment:
I think with the new host header check we now need two more tests:
- When disable domain fronting is true / mtls, it sends a 400 bad request on missing host header
- When disable domain fronting is false, it sends a 400 bad request on missing host header (this a regression test that makes sure the default behavior is unchanged)

Reply:
will do this as a separate PR

Reply:
Resolving: As this requires a TCP-level connection without using the http client lib, it might be more complicated. Agreed to cut a release first, then add the test.